File: libs/ts/src/calendars.ts
/**
 * Calendars — per-year file loader.
 * Loads year JSON files from a directory path or base URL.
 */

import { BellSchedule } from './bell-schedule.js';
import type { BellScheduleOptions, YearData } from './types.js';

class Calendars {
  #basePath: string;
  #cache: Map<string, YearData[]>;

  /**
   * @param basePath - Directory path (e.g. './calendars/') or URL base.
   */
  constructor(basePath: string) {
    this.#basePath = basePath;
    this.#cache = new Map(); // year string → parsed data
  }

  /** Load data for a specific academic year (e.g. '2025-2026'). */
  async #load(year: string): Promise<YearData[]> {
    const cached = this.#cache.get(year);
    if (cached) {
      return cached;
    }

    const filePath = `${this.#basePath}${year}.json`;
    let data: unknown;

    if (this.#basePath.startsWith('http://') || this.#basePath.startsWith('https://')) {
      const res = await fetch(filePath);
      if (!res.ok) throw new Error(`Failed to fetch ${filePath}: ${res.status} ${res.statusText}`);
      data = await res.json();
    } else {
      const { readFile } = await import('node:fs/promises');
      const text = await readFile(filePath, 'utf8');
      data = JSON.parse(text);
    }

    // Normalize to array.
    const arr = (Array.isArray(data) ? data : [data]) as YearData[];
    this.#cache.set(year, arr);
    return arr;
  }

  /** Get a BellSchedule for a specific academic year (e.g. '2025-2026'). */
  async forYear(year: string, options: BellScheduleOptions = {}): Promise<BellSchedule> {
    const arr = await this.#load(year);
    return new BellSchedule(arr, options);
  }

  /**
   * Get a BellSchedule appropriate for the current instant.
   * During summer, loads both the most recent ended year and the next upcoming
   * year so summer-bounds and next-year-start queries work correctly.
   *
   * "Today" defaults to the system-local date; pass `timeZone` to anchor the
   * academic-year rollover to a specific zone (e.g. the school's) when running
   * elsewhere — e.g. a server in UTC.
   */
  async current(options: BellScheduleOptions = {}, timeZone?: string): Promise<BellSchedule> {
    const today = Temporal.Now.plainDateISO(timeZone);
    const year = this.#academicYearFor(today);

    const primaryArr = await this.#load(year);
    const firstDay = primaryArr[0].firstDayTeachers || primaryArr[0].firstDay;
    const lastDay = primaryArr[0].lastDay;
    const inYear = today.toString() >= firstDay && today.toString() <= lastDay;

    if (inYear) {
      return new BellSchedule(primaryArr, options);
    }

    // Summer — load adjacent year.
    const allData = [...primaryArr];

    if (today.toString() > lastDay) {
      // After this year's end — load the next academic year.
      const nextYearLabel = this.#nextAcademicYear(year);
      try {
        const nextArr = await this.#load(nextYearLabel);
        allData.push(...nextArr);
      } catch {
        // Next year data not available; that's fine.
      }
    } else {
      // Before this year's start — load the previous academic year.
      const prevYearLabel = this.#prevAcademicYear(year);
      try {
        const prevArr = await this.#load(prevYearLabel);
        allData.unshift(...prevArr);
      } catch {
        // Previous year data not available; that's fine.
      }
    }

    return new BellSchedule(allData, options);
  }

  /**
   * Determine the academic year label for a given date.
   * Academic year starts in August.
   */
  #academicYearFor(date: Temporal.PlainDate): string {
    const { month, year } = date;
    if (month >= 8) {
      return `${year}-${year + 1}`;
    } else {
      return `${year - 1}-${year}`;
    }
  }

  #nextAcademicYear(year: string): string {
    const [start] = year.split('-').map(Number);
    return `${start + 1}-${start + 2}`;
  }

  #prevAcademicYear(year: string): string {
    const [start] = year.split('-').map(Number);
    return `${start - 1}-${start}`;
  }
}

export { Calendars };

File: libs/python/bells/calendars.py
"""Calendars — per-year file loader.

Python counterpart of the JavaScript ``calendars.js`` module. Loads year JSON
files from a directory path or a base URL.
"""

from __future__ import annotations

import json
from datetime import date
from pathlib import Path
from typing import Optional
from urllib.request import urlopen

from .bell_schedule import BellSchedule
from .datetimeutil import _instant_to_date, _now_instant


class Calendars:
    """Loads per-year calendar JSON files lazily and builds BellSchedules."""

    def __init__(self, base_path: str):
        """``base_path`` is a directory path (e.g. ``'./calendars/'``) or URL base."""
        self._base_path = base_path
        self._cache: dict[str, list[dict]] = {}

    def _load(self, year: str) -> list[dict]:
        if year in self._cache:
            return self._cache[year]

        file_path = f"{self._base_path}{year}.json"

        if self._base_path.startswith("http://") or self._base_path.startswith("https://"):
            with urlopen(file_path) as res:
                data = json.loads(res.read().decode("utf-8"))
        else:
            text = Path(file_path).read_text(encoding="utf-8")
            data = json.loads(text)

        arr = data if isinstance(data, list) else [data]
        self._cache[year] = arr
        return arr

    def for_year(self, year: str, options: Optional[dict] = None) -> BellSchedule:
        """Build a BellSchedule for a specific academic year (e.g. ``'2025-2026'``)."""
        arr = self._load(year)
        return BellSchedule(arr, options or {})

    def current(self, options: Optional[dict] = None, tz: Optional[str] = None) -> BellSchedule:
        """Build a BellSchedule appropriate for the current date.

        During summer, loads both the most recent ended year and the next
        upcoming year so summer-bounds and next-year-start queries work.

        "Today" defaults to the system-local date; pass ``tz`` (an IANA name) to
        anchor the academic-year rollover to a specific zone (e.g. the school's)
        when running elsewhere — e.g. a server in UTC.
        """
        today = _instant_to_date(_now_instant(), tz) if tz else date.today()
        year = self._academic_year_for(today)

        primary_arr = self._load(year)
        first_day = primary_arr[0].get("firstDayTeachers") or primary_arr[0]["firstDay"]
        last_day = primary_arr[0]["lastDay"]
        today_str = today.isoformat()
        in_year = first_day <= today_str <= last_day

        if in_year:
            return BellSchedule(primary_arr, options or {})

        # Summer — load adjacent year.
        all_data = list(primary_arr)

        if today_str > last_day:
            # After this year's end — load the next academic year.
            next_year_label = self._next_academic_year(year)
            try:
                all_data.extend(self._load(next_year_label))
            except Exception:
                pass  # Next year data not available; that's fine.
        else:
            # Before this year's start — load the previous academic year.
            prev_year_label = self._prev_academic_year(year)
            try:
                all_data[:0] = self._load(prev_year_label)
            except Exception:
                pass  # Previous year data not available; that's fine.

        return BellSchedule(all_data, options or {})

    def _academic_year_for(self, d: date) -> str:
        """Academic year label for a date. The academic year starts in August."""
        month, year = d.month, d.year
        if month >= 8:
            return f"{year}-{year + 1}"
        return f"{year - 1}-{year}"

    def _next_academic_year(self, year: str) -> str:
        start = int(year.split("-")[0])
        return f"{start + 1}-{start + 2}"

    def _prev_academic_year(self, year: str) -> str:
        start = int(year.split("-")[0])
        return f"{start - 1}-{start}"
